DRILL-7828: Refactor Pcap and Pcapng format plugin

vdiravka opened the pull request:
# [DRILL-7828](https://issues.apache.org/jira/browse/DRILL-7828): Refactor Pcap and Pcapng format plugin

## Description

It merges two (PCAP and PCAPNG) format plugins into one and placed it in the `contrib` module. Hence it uses two different readers for them. The decision, which reader to use is based on identifying PCAP file Magic Number

## Documentation
An example of setting PCAP Format Plugin
```
        "pcap": {
          "type": "pcap",
          "extensions": ["pcap", "pcapng"],
          "stat" : false, // optional
          "sessionizeTCPStreams": false // optional
        }
```

## Testing
Added test case for the `pcapng` file, but `pcap` extension

luocooong commented:
@vdiravka Thanks for the nice submission. I'll do a review on this weekend.
BTW, Has this PR implemented the `Streaming read for Pcap-NG format` feature?

paul-rogers commented on contrib/format-pcapng/src/main/java/org/apache/drill/exec/store/plugin/PcapFormatConfig.java:
```diff
@@ -29,18 +29,23 @@
-    this.extensions = extensions == null ? ImmutableList.of(PcapngFormatConfig.NAME) : ImmutableList.copyOf(extensions);
+  public PcapFormatConfig(@JsonProperty("extensions") List<String> extensions,
+                          @JsonProperty("stat") boolean stat,
+                          @JsonProperty("sessionizeTCPStreams") Boolean sessionizeTCPStreams) {
```
People have to type this name. Brevity is good. Maybe just `sessionize` as the property name. Because this is PCAP, the only thing that *can* be sessionized are TCP streams.

paul-rogers replied:
Thanks for the explanation. If the property already existed, then we must leave it to preserve backward compatibility. I guess we should have caught the issue when the property was originally added.

paul-rogers commented on contrib/format-pcapng/src/main/java/org/apache/drill/exec/store/plugin/PcapFormatPlugin.java:
```diff
@@ -0,0 +1,157 @@
+import org.apache.drill.common.types.TypeProtos.MinorType;
+import org.apache.drill.common.types.Types;
+import org.apache.drill.exec.physical.impl.scan.file.FileScanFramework;
+import org.apache.drill.exec.physical.impl.scan.file.FileScanFramework.FileReaderFactory;
```
At some point it would be good to start using the "new" version of this stuff which is considerably simpler, and more forward looking, than this original version. For example, the "new" version provides better options for working with provided schemas.

paul-rogers replied:
We don't have a ticket (that I filed). I got as far as doing the framework and updating CSV. My thought was, if we are refactoring a plugin, might as well use the newest stuff. But, it is fine to do that some other time if/when there is a compelling reason to do so.

paul-rogers commented on contrib/format-pcapng/src/main/java/org/apache/drill/exec/store/plugin/PcapFormatPlugin.java:
```diff
@@ -0,0 +1,157 @@
+
+  /**
+   * Detects PCAP or PCAPNG file format based on file extension. Not used now due to {@link #fromMagicNumber} usage for
+   * this purpose
```
Drill conventions have been to remove unused code.

paul-rogers commented on contrib/format-pcapng/src/main/resources/bootstrap-format-plugins.json:
```diff
@@ -3,26 +3,29 @@
-        "pcapng": {
-          "type": "pcapng",
+        "pcap": {
+          "type": "pcap",
```
Breaking change; see above.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/scan/columns/ColumnsScanFramework.java:
```diff
@@ -103,7 +103,7 @@ protected void configure() {
 
   @Override
-  protected SchemaNegotiatorImpl newNegotiator() {
+  public SchemaNegotiatorImpl newNegotiator() {
```
Why is this now `public`? The `columns` framework was for CSV files using the `columns` column; this is an internal bit of structure not meant to be overridden.

Also, would be better to invest time with the "new" version rather than this somewhat cumbersome older version.

paul-rogers replied:
As it turns out, cleaning up this kind of issue is one of the things that the new version does.

There is also a race condition. The schema negotiator is passed into the reader; it is not meant to be obtained elsewhere. In general, the framework does a bunch of initialization, getting things ready. Then, each reader is created, one at a time, passing in the schema negotiator. If we have code trying to access it at some other time, then we'd have to review all that code to see if new pattern works.

vdiravka replied:
Took this into account with new method:
```
    /**
     * @return FileScanFramework or empty object in case it is not binded yet with {@link #bind(ManagedScanFramework)}
     */
    protected Optional<FileScanFramework> fileFramework() {
      return Optional.ofNullable(fileFramework);
    }
```
and
```
It is not possible to detect file format, because the File Framework is not initialized. " +
                "Trying to use default PCAP format plugin to read the file
```
It will help for future new version of `FileReaderFactory`

vdiravka reviewed (commented):
@luocooong there is no changes in Pcap-NG reader in this PR. This PR only merges two plugins into one and add proper format identifying based on PCAP magic number, but two old readers still used. And they can be improved in further work.

@paul-rogers thanks for CR. Please check my answers

paul-rogers commented on contrib/format-pcapng/src/main/java/org/apache/drill/exec/store/pcap/PcapBatchReader.java:
```diff
@@ -48,126 +49,61 @@
+  public PcapBatchReader(PcapFormatConfig readerConfig, int maxRecords) {
     this.readerConfig = readerConfig;
-    if (readerConfig.sessionizeTCPStreams) {
+    if (readerConfig.getSessionizeTCPStreams()) {
```
Good question. I think the issue is tied up with table functions. Try using your `private` members in a query that uses table functions to set plugin properties. This is a bit hazy now, but we changed something about that a while back to make things more general.

In fact, a good rule of thumb would be to always add a table function unit test for each plugin and its properties.

vdiravka replied:
I'll add:
```
SELECT type, packet_length, `timestamp` FROM table(dfs.`pcapng/sniff.pcapng` (type => 'pcapng', stat => false, sessionizeTCPStreams => true )) where type = 'ARP
```

paul-rogers commented on contrib/format-pcapng/src/main/java/org/apache/drill/exec/store/plugin/PcapFormatConfig.java:
```diff
@@ -29,18 +29,23 @@
+  private static final List<String> DEFAULT_EXTNS = ImmutableList.of("pcap", "pcapng");
 
-  public static final String NAME = "pcapng";
+  public static final String NAME = "pcap";
```
There may be a trick to test this. In some of the storage plugin rework we did, we added some tests to push things into the persistent store. You might be able to look at that and write an old-style config to the store as JSON, then try to retrieve it in a query.

vdiravka replied:
I did it :)
Please check `TestPcapWithPersistentStore#pcapPluginBackwardCompatabilityTest()`. It is written similar to `TestPStoreProviders#zkBackwardCompatabilityTest()`.
It serializes JSON file to `FileSystemConfig` object, put it to ZK PersistentStore and retrieves it from there as `PcapFormatConfig` and `PcapngFormatConfig` types.

paul-rogers reviewed (commented):
@luocooong, thanks for your note. How familiar are you with how Drill handles storage plugins? It is far more fragile than we'd like. Suppose a user uses the "obsolete" plugin and thus creates a config in ZK. Now, suppose we remove that old plugin, the user upgrades Drill, and starts the new version? What happens? Historically, Drill fails to start with a very obscure message. Why? Because the storage plugin registry read the old JSON, could not find the corresponding Java class, and failed deep in Jackson, which causes Drill to fail to start. A real mess. I got bit by this early in my Drill days; it took a while to figure it out and I only did because another person had recently hit the same issue.

We revised the plugin registry last year, but we didn't (IIRC) fix this issue. We could do so, but I think I saved that for a later round of improvements (which I didn't ever get to.)

So, this is why we have to be super-careful when changing plugin configs: we have to preserve the very brittle plugin compatibility rules.

paul-rogers commented:
One other thought on the compatibility issue. We can perhaps fudge it here as we discussed. Perhaps it make sense to fix it correctly:

* Store a version number in the "persistent store" (ZK) each time we write to the store. This can't be a product version, it must be a persistent store format version. (Why? So you can change things in the SNAPSHOT release.)
* On each startup, check the persistent store version. If it is less than the current version, run an upgrade process. In particular, invoke a `versionxToy` conversion for each 'x' from the current store version to the current software version.
* In each conversion, look for changed configs and "do the right thing." For this PR, it would mean changing the plugin name and rewriting one of them ("pcap-ng" maybe) as needed. In some cases, it might mean deleting an unsupported plugin. Etc.

The wise thing would be to log the original JSON and any changes, so if things go wrong (and they will), there is a way to recover.

luocooong commented:
> One other thought on the compatibility issue. We can perhaps fudge it here as we discussed. Perhaps it make sense to fix it correctly:
> 
> * Store a version number in the "persistent store" (ZK) each time we write to the store. This can't be a product version, it must be a persistent store format version. (Why? So you can change things in the SNAPSHOT release.)
> * On each startup, check the persistent store version. If it is less than the current version, run an upgrade process. In particular, invoke a `versionxToy` conversion for each 'x' from the current store version to the current software version.
> * In each conversion, look for changed configs and "do the right thing." For this PR, it would mean changing the plugin name and rewriting one of them ("pcap-ng" maybe) as needed. In some cases, it might mean deleting an unsupported plugin. Etc.
> 
> The wise thing would be to log the original JSON and any changes, so if things go wrong (and they will), there is a way to recover.

Nice Idea. Versioning + Upgrade from previous + Recovery mode

vdiravka commented:
@paul-rogers I am familiar with this issue in Drill:
https://stackoverflow.com/questions/50487014/apache-drill-1-12-0-failure-in-starting-embedded-drillbit-and-no-current-conn/50495562#50495562

And I am absolutely agree about plugins configs versioning. We have it for parquet metdata, smth similar can be done for plugins. And other thing is proper migration tool for the new schema of storage plugins configs (every plugin configs change will require to update plugins configs version and migration on the new Drill sturtup with an old configs).

But it sounds like a useful, but new feature. It shouldn't prevent to go further with current PR. Especially in current PR, I added code for an old plugins configs, so new Drill start-up will not cause any issues with an old plugin configs.

luocooong commented:
Actually, There is a JIRA (thanks @vdiravka) related to this ideas. [Storage Plugins Handler improvements](https://issues.apache.org/jira/browse/DRILL-6564)
Well,  If this feature is implemented ASAP, it will be great. (1.19 or later ?)

luocooong commented:
@vdiravka Hi. Do we have any new revisions ready to submit? AFAIK, Now there is only an issue and related to the `conflict upragde`, Is right?

paul-rogers commented:
Where are we on this one? We should avoid the "boil the ocean" temptation: the desire to fix too many things in one PR. If this PR will break upgrades, then the team can decide if that is OK for the handful of people who might be affected. Maybe document recovery in release notes, or be prepared to describe the steps on Slack or the mail lists.

I am concerned about table functions. They are obscure, but are supposed to work. I'd like to see a test added so we can be sure that the non-public config variables do, in fact, work. Of course, you could argue that we can skip that until someone complains, then file another bug because, again, there are likely few users of this plugin and likely none of them use table functions.

luocooong commented:
@dzamo Hello. Is it possible to add a docs (when the [DRILL-6564](https://issues.apache.org/jira/browse/DRILL-6564) or related issues not resolved) that describe the troubleshooting about `conflict upgrade` at the phase of upgrade (from old plugin config). thanks.

vdiravka commented:
As I told earlier we don't break backward compatibility: in second commit `Changes related to CR` I have added support for old plugins and in third commit `Changes related to CR2` I have added test case to check it. So we are fine to use the new config and don't fail, if old plugin config is places in PS
@paul-rogers Could you pls check again?

paul-rogers reviewed (approved):
Thanks for the great work. LGTM.

+1

luocooong commented:
@paul-rogers Thanks for your review and thanks for your time.

luocooong reviewed (approved):
@vdiravka Thanks for your PR and revisions. Please squash all the commits.
